magma/types/ready_valid.py

```diff
@@ -29,14 +29,25 @@ def _check_T(cls, T):
             )
         return undirected_T
 
+    def _add_data_if_not_none(cls, fields, T, dir):
+        if T is not None:
+            T = cls._check_T(T)
+            fields["data"] = dir(T)
```

Comment:
```suggestion
    def _maybe_add_data(cls, fields, T, qualifier):
        if T is None:
            return
        T = cls._check_T(T)
        fields["data"] = qualifier(T)
```

and let's make it a free function and then invoke it as `_maybe_add_data(cls, ...)`

Reply:
Actually on second thought we can probably have a single free function:

```python
_make_fields(cls, T, qualifier):
    fields = {"valid": Bit, "ready": Bit}
    if T is None:
        return fields
    T = cls._check_T
    fields["data"] = qualifier(T)
    return fields
```

on that note `_check_T` can also be free

(my general principle for free functions is if we don't want them to be public *and* they're "stateless" [i.e. do not modify state of the class/instance] then they can be free.)

Reply:
We can make it a free function, but we can't reuse _make_fields the way you wrote it because the producer/consumer kind use different directions for valid/ready, so we either need to add arguments for those qualifiers, but at that point I think it's about the same amount of code as just passing in the initialized dictionary